File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-5.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    goodNodes(root) {        
        function dfs(node, maxSoFar){
            if (!node) return 0;

            let count = 0

            // curr node is good, cause its >= maxSoFar
            if (node.val >= maxSoFar){
                count = 1;
            }

            let newMax = Math.max(maxSoFar, node.val);

            // add good nodes from the left and right side
            count += dfs(node.left, newMax);
            count += dfs(node.right, newMax);

            return count;
        }

        return dfs(root, root.val);
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0

        def dfs(node):
            nonlocal res

            if not node:
                return 0

            right = 0
            left = 0

            if node.right:
                right = dfs(node.right)

            if node.left:
                left = dfs(node.left)

            res = max(res, right + left)
            return 1 + max(right, left)

        dfs(root)
        return res